src/Microsoft.AspNetCore.Razor.VSCode/src/RazorDocumentSynchronizer.ts

```diff
@@ -24,164 +25,257 @@ export class RazorDocumentSynchronizer {
             }
-            synchronizationContext.synchronized(false);
-        }, 500);
-        synchronizationContext = {
-            logIdentifier,
-            toVersion,
-            documentVersion: hostDocument.version,
-            synchronized: (s) => {
-                delete this.synchronizations[uriPath];
-                clearTimeout(timeout);
-                synchronized(s);
-            },
-            onSynchronized,
-        };
-        this.synchronizations[uriPath] = synchronizationContext;
 
-        const success = await onSynchronized;
+            // Projected text document is the one we expect
+        } catch (cancellationReason) {
+            this.removeSynchronization(context);
 
-        if (success && projectedDocument.hostDocumentSyncVersion !== hostDocument.version) {
             if (this.logger.verboseEnabled) {
-                this.logger.logVerbose(`${logIdentifier} - Failed: User moved on.`);
+                this.logger.logVerbose(
+                    `${logId} - Synchronization failed: ${cancellationReason}`);
             }
 
-            // Already out-of-date, failed to synchronize.
             return false;
         }
 
-        if (success) {
-            if (this.logger.verboseEnabled) {
-                this.logger.logVerbose(`${logIdentifier} - Success: Documents synchronized to version ${toVersion}.`);
-            }
+        this.removeSynchronization(context);
 
-            return true;
-        } else {
-            if (this.logger.verboseEnabled) {
-                this.logger.logVerbose(`${logIdentifier} - Failed: Documents not synchronized ${toVersion}.`);
+        if (this.logger.verboseEnabled) {
+            this.logger.logVerbose(
+                `${logId} - Synchronization successful!`);
+        }
+
+        return true;
+    }
+
+    private removeSynchronization(context: SynchronizationContext) {
+        const documentKey = getUriPath(context.projectedDocument.uri);
+        const synchronizations = this.synchronizations[documentKey];
+        clearTimeout(context.timeoutId);
+
+        if (synchronizations.length === 1) {
+            delete this.synchronizations[documentKey];
+            return;
+        }
+
+        for (let i = 0; i < synchronizations.length; i++) {
+            if (synchronizations[i] === context) {
+                const newSynchronizations = synchronizations.splice(i, 1);
+                this.synchronizations[documentKey] = newSynchronizations;
+                break;
             }
+        }
+    }
 
-            return false;
+    private createSynchronizationContext(
+        documentKey: string,
+        projectedDocument: IProjectedDocument,
+        toHostDocumentVersion: number,
+        hostDocument: vscode.TextDocument,
+        token: vscode.CancellationToken) {
+
+        const rejectionsForCancel: Array<(reason: string) => void> = [];
+        let projectedDocumentSynchronized: () => void = Function;
```

Comment:
This is interesting. I've never seen the `Function` object used as an actual function instance before.

Is this definitely what you want? I just tried it, and at least in browsers, attempting to invoke `Function()` triggers a CSP exception. Maybe it's OK in VS Code's JS runtime. Perhaps it would be more conventional to replace this with:

```javascript
let projectedDocumentSynchronized = () => {};
```

... if you definitely need a no-op. I know that will allocate more but so is everything else.

Reply:
Oh it looks like you're not using the initial value anyway. So it could be:

```javascript
let projectedDocumentSynchronized: Function;
```

Reply:
Have to assign it to something because:
![image](https://user-images.githubusercontent.com/2008729/48283207-af82e180-e410-11e8-838d-1c74ac69fa2e.png)

And tslint complains about using `Function` as a type:
![image](https://user-images.githubusercontent.com/2008729/48283250-cf1a0a00-e410-11e8-8a0e-86d4b7385f39.png)

Either way the variable immediately gets set in the promise creation. I was doing this with the intent of pre-setting a variable to `null` like in C#. There a better way to work around the limitations in  clean way?

Reply:
Setting it to `null` would be fine (`null` !== `Function`).

Reply:
Would have to null check it before invoking it then :/. Gonna leave it as Function for now, it'll be immediately overridden anyhow.